## Final Project Ai Algorithms - Tom Delahaye Gabriel Carlotti - DIA 3 🏧

This notebook focuses on solving a biscuit optimization problem where biscuits of various types need to be placed on a dough strip. The goal is to maximize the total profit by selecting the best arrangement of biscuits while adhering to specific constraints:

- Non-overlapping biscuits: No two biscuits can occupy the same position on the dough.
- Defect tolerances: Each biscuit type can only tolerate a certain number of defects within its covered positions.
- Penalty for unused positions: Any unused position on the dough incurs a penalty to the overall score.

# 0. Imports & basics preparation

In this section, we prepare the necessary components for the biscuit optimization problem. The main goal is to initialize the data structures and objects that will represent the problem's constraints and variables.

Importing Modules:

Essential libraries and custom modules (Biscuit, Dough) are imported to handle data, model biscuits, and implement optimization algorithms.
Loading Defects Data:

A CSV file defects.csv containing defect information is loaded into a Pandas DataFrame. Each defect is characterized by its position on the dough and its type ('a', 'b', or 'c').
Creating Biscuit Objects:

Five biscuit types are instantiated using the Biscuit class. Each biscuit type has unique properties, including:
- Length: How many units of the dough the biscuit covers.
- Value: The profit associated with placing the biscuit.
- Defect Tolerance: The maximum allowable defects of each type ('a', 'b', 'c') within the biscuit's range.

Classes Overview:

- Biscuit Class:
Represents a specific type of biscuit with predefined properties.
Configures each biscuit's length, value, and defect tolerance based on its type.

- Dough Class:
Represents the dough as a linear strip of a given length.
Manages defects, validates biscuit placements, and calculates the total value of placements.
Includes methods to check overlapping biscuits, calculate penalties for unused positions.

In [1]:
import pandas as pd
import random
import time
from Biscuit import *
from Dough import *
from GeneticAlgorithm import *
from GeneticElitism import * 
from GeneticTournament import *
from UniformCrossoverGA import *

In [2]:
df = pd.read_csv('defects.csv', sep =',')

In [3]:
df.head()

x class
0  355.449335     c
1   92.496236     a
2  141.876795     c
3  431.833902     c
4  435.028461     c

In [4]:
biscuit0 = Biscuit(0)
biscuit1 = Biscuit(1)
biscuit2 = Biscuit(2)
biscuit3 = Biscuit(3)
biscuit4 = Biscuit(4)

print(biscuit0)
print(biscuit1)
print(biscuit2)
print(biscuit3)
print(biscuit4)

Biscuit Type: 0, Length: 4, Value: 3, Max Defects: {'a': 4, 'b': 2, 'c': 3}
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}
Biscuit Type: 2, Length: 2, Value: 1, Max Defects: {'a': 1, 'b': 2, 'c': 1}
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}
Biscuit Type: 4, Length: 1, Value: -1, Max Defects: {'a': 10, 'b': 10, 'c': 10}


In [5]:
dough = Dough(length=500)

for _, row in df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])

In [6]:
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

# 1. Genetic Algorithm

## 1.1 Basic Genetic Algorithm

In [7]:
#Simple program to launch the Genetic Algorithm
def main():
    dough = Dough(length=500)
    for _, row in df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
        
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    GA = GeneticAlgorithm(dough, biscuits, population_size=50, mutation_rate=0.01, crossover_rate=0.7)
    GA.initialize_population()

    for generation in range(100): 
        print(generation)
        GA.evolve()

    best_solution_basic = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution_basic)
    print("Total Value:", total_value)
    print("Best Solution:", best_solution_basic)
    
    return best_solution_basic


- Population size = 50: Chosen to balance exploration and computational overhead.
- Mutation rate = 0.01: A low mutation rate to introduce variability without destabilizing solutions.
- Crossover rate = 0.7: Ensures effective combination of parent solutions to explore the solution space.
- Generation = 100 : Evolves the population over 100 generations, providing sufficient iterations to converge to a high-quality solution.

In [8]:
if __name__ == "__main__":
    start = time.time()
    best_solution_basic = main()
    print("solved in : ", time.time()-start,"seconds")

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
Total Value: 613
Best Solution: [(0, 1), (8, 1), (16, 1), (24, 3), (29, 3), (34, 2), (36, 0), (40, 0), (44, 0), (48, 2), (50, 0), (54, 1), (62, 3), (67, 2), (69, 3), (74, 0), (78, 3), (83, 1), (91, 2), (93, 0), (98, 1), (106, 2), (108, 2), (111, 0), (118, 0), (122, 1), (130, 1), (138, 0), (142, 0), (146, 2), (148, 3), (153, 2), (155, 2), (158, 1), (167, 3), (172, 2), (174, 0), (178, 0), (182, 2), (185, 2), (187, 1), (195, 1), (203, 2), (205, 3), (210, 1), (218, 3), (223, 1), (231, 3), (236, 0), (242, 1), (250, 3), (255, 1), (263, 1), (271, 3), (276, 1), (284, 3), (289, 2), (291, 2), (293, 1), (301, 1), (309, 1), (317, 3), (322, 1), (330, 0), (334, 3), (339, 1), (347, 1), (355, 1), (363, 0), (367, 0), 

In [9]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

In [10]:
dough = Dough(length=500)
for _, row in df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
    
dough.place_biscuits_GA(best_solution_basic)
dough.calculate_own_value(biscuits_info)
dough.validate_biscuits_placement(biscuits_info)

(True, 'All biscuits are correctly placed and within defect thresholds.')

The output value is preetty low, but that could be expected since our population is pretty small (50) and the number of defects far exceeds that, we would need a bigger population to reinforce our results or increase our mutation rates to find by hazard a better solution

## 1.2 Genetic Algorithm with Elitism Roulette

In [ ]:
def main():
    dough = Dough(length=500) 
    for _, row in df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    GA = GeneticElitism(dough, biscuits, population_size=50, mutation_rate=0.01, crossover_rate=0.7)
    GA.initialize_population()

    for generation in range(100): 
        print(generation)
        GA.evolve()

    # Find the best solution in the final population
    best_solution_elitism = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution_elitism)
    print("Total Value:", total_value)
    print("Best Solution:", best_solution_elitism)
    return best_solution_elitism

In [12]:
if __name__ == "__main__":
    start = time.time()
    best_solution_elitism = main()
    print("solved in : ", time.time()-start,"seconds")

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
Total Value: 630
Best Solution: [(0, 1), (8, 1), (16, 3), (21, 2), (23, 1), (31, 3), (36, 0), (40, 1), (48, 2), (50, 0), (54, 1), (62, 3), (67, 3), (72, 2), (74, 1), (82, 1), (90, 2), (92, 0), (98, 1), (106, 0), (111, 0), (118, 0), (122, 2), (124, 1), (132, 3), (137, 2), (139, 2), (141, 0), (145, 0), (149, 3), (154, 2), (158, 0), (162, 0), (167, 3), (172, 0), (176, 1), (185, 2), (187, 1), (195, 3), (200, 3), (205, 3), (210, 3), (215, 3), (220, 3), (225, 3), (230, 3), (235, 2), (237, 1), (245, 0), (249, 3), (254, 3), (259, 0), (263, 3), (268, 3), (273, 0), (277, 1), (285, 3), (290, 2), (293, 1), (301, 1), (309, 3), (314, 1), (322, 3), (327, 0), (331, 2), (333, 1), (341, 1), (349, 2), (351, 2), (353, 0)

In [39]:
dough = Dough(length=500)
for _, row in df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
    
dough.place_biscuits_GA(best_solution_elitism)
print(dough.calculate_own_value(biscuits_info))
dough.validate_biscuits_placement(biscuits_info)

630


(True, 'All biscuits are correctly placed and within defect thresholds.')

The implementation of the Genetic Algorithm with elitism has yielded an improved result of 630, outperforming the previous value of 613. This improvement highlights the effectiveness of elitism, where the top-performing individuals are preserved across generations, ensuring that their quality traits are retained. By reserving 20% of the population as elite, the algorithm avoids the risk of losing high-quality solutions due to random mutations or suboptimal crossovers. The combination of elitism, roulette wheel selection, and mutation introduces a balanced exploration and exploitation of the solution space, leading to a higher-quality placement of biscuits on the dough while respecting all constraints.

## 1.3 Genetic Algorithm with Tournament Selection and Crossover

In [14]:
def main():
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }
    dough = Dough(500)

    for _, row in df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])

    population_size = 150
    mutation_rate = 0.1
    crossover_rate = 0.5
    GA = GeneticTournament(dough, biscuits, population_size, mutation_rate, crossover_rate)
    num_generations = 100
    
    for generation in range(num_generations):
        print(generation)
        GA.evolve()

    best_solution_tournament = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution_tournament)
    print("Best Solution Total Value:", total_value)
    print("Best Solution Configuration:", best_solution_tournament)
    return best_solution_tournament


In [15]:
if __name__ == "__main__":
    start = time.time()
    best_solution_tournament = main()
    print("solved in : ", time.time()-start,"seconds")

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
Best Solution Total Value: 639
Best Solution Configuration: [(0, 1), (462, 1), (8, 1), (414, 1), (16, 1), (24, 3), (167, 3), (29, 3), (238, 2), (34, 2), (143, 0), (422, 0), (122, 1), (48, 2), (50, 0), (224, 1), (54, 1), (118, 0), (62, 1), (381, 2), (368, 0), (134, 3), (78, 3), (152, 3), (83, 0), (104, 2), (87, 3), (398, 1), (92, 0), (497, 2), (98, 0), (102, 2), (242, 1), (325, 3), (106, 2), (108, 3), (284, 0), (362, 0), (487, 3), (346, 1), (258, 1), (40, 1), (139, 0), (406, 1), (158, 1), (187, 1), (172, 0), (470, 3), (176, 1), (185, 2), (446, 1), (211, 1), (195, 1), (203, 1), (317, 1), (219, 3), (232, 0), (266, 3), (236, 2), (438, 1), (114, 2), (250, 1), (434, 0), (271, 3), (373, 1), (276, 1), (492, 3

In [40]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

dough = Dough(length=500)
for _, row in df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
    
dough.place_biscuits_GA(best_solution_tournament)
print(dough.calculate_own_value(biscuits_info))
dough.validate_biscuits_placement(biscuits_info)

639


(True, 'All biscuits are correctly placed and within defect thresholds.')

The implementation of tournament selection combined with crossover and 20% elitism resulted in a total value of 639, an improvement over the previous value of 630. Tournament selection enhances the algorithm by focusing on a subset of individuals in each selection round, favoring fitter candidates while still allowing diversity. This method, coupled with crossover, ensures effective exploration of the solution space by combining traits from selected parents. The 20% elitism guarantees the preservation of top-performing individuals across generations, maintaining a high-quality baseline. These mechanisms work together to achieve a balance between exploiting known good solutions and exploring new ones, yielding better results within the same number of generations.

## 1.4 Genetic Algorithm with Uniform Crossover

In [42]:
def main():
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }
    dough = Dough(500)

    for _, row in df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])

    population_size = 150
    mutation_rate = 0.1
    crossover_rate = 0.5
    GA = UniformCrossoverGA(dough, biscuits, population_size, mutation_rate, crossover_rate)
    num_generations = 100
    
    for generation in range(num_generations):
        print(generation)
        GA.evolve()

    best_solution_uniform = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution_uniform)
    print("Best Solution Total Value:", total_value)
    print("Best Solution Configuration:", best_solution_uniform)
    
    return best_solution_uniform

In [43]:
if __name__ == "__main__":
    start = time.time()
    best_solution_uniform = main()
    print("solved in : ", time.time()-start,"seconds")

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
Best Solution Total Value: 640
Best Solution Configuration: [(0, 1), (8, 1), (16, 3), (21, 1), (29, 3), (34, 2), (36, 0), (40, 1), (48, 2), (50, 0), (54, 3), (59, 1), (67, 3), (72, 1), (80, 1), (88, 1), (98, 0), (102, 2), (104, 1), (112, 0), (118, 0), (122, 2), (124, 1), (132, 2), (134, 3), (139, 0), (143, 0), (147, 0), (151, 3), (158, 1), (167, 3), (172, 2), (174, 1), (182, 2), (185, 2), (187, 1), (195, 3), (200, 3), (205, 0), (209, 3), (214, 0), (218, 1), (226, 1), (234, 2), (236, 3), (242, 1), (250, 1), (258, 0), (262, 3), (267, 3), (272, 2), (274, 0), (278, 2), (280, 1), (288, 3), (293, 1), (301, 1), (309, 1), (317, 1), (325, 1), (333, 3), (338, 3), (343, 3), (348, 3), (353, 0), (357, 3), (362, 0)

In [44]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

dough = Dough(500)
for _, row in df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
dough.place_biscuits_GA(best_solution_uniform)
print(dough.calculate_own_value(biscuits_info))
dough.validate_biscuits_placement(biscuits_info)

640


(True, 'All biscuits are correctly placed and within defect thresholds.')

Using uniform crossover with tournament selection and elitism resulted in a total value of 640, which stays an improvement over the previous value of 639. This version introduces two key changes to the algorithm parameters and methodology:

- Tournament Size: Increased to 15, allowing for a more competitive selection process. This ensures that only the most promising candidates among a larger pool are chosen for crossover, thereby improving the average fitness of the selected individuals.

- Uniform Crossover: Unlike traditional one-point or two-point crossover, uniform crossover allows for gene-level mixing between parents. This method provides more diverse offspring by selecting genes independently from either parent, leading to a better exploration of the solution space. The use of valid crossover points prevents overlapping and ensures feasibility in offspring solutions.


# 2. Greedy Search

## 2.1 Finding Heuristics

In [20]:
biscuits = [biscuit0,biscuit1,biscuit2,biscuit3]

for biscuit in biscuits:
    print("Biscuit type: ",biscuit.biscuit_type)
    print("Value/Length: ",biscuit.value/biscuit.length)
    maxd = sum(biscuit.max_defects.values())
    print("Max defects: ", maxd)
    print("Value/(Length*Max_Defects): ", biscuit.value/(biscuit.length*maxd), "\n")

Biscuit type:  0
Value/Length:  0.75
Max defects:  9
Value/(Length*Max_Defects):  0.08333333333333333 

Biscuit type:  1
Value/Length:  1.5
Max defects:  13
Value/(Length*Max_Defects):  0.11538461538461539 

Biscuit type:  2
Value/Length:  0.5
Max defects:  4
Value/(Length*Max_Defects):  0.125 

Biscuit type:  3
Value/Length:  1.6
Max defects:  7
Value/(Length*Max_Defects):  0.22857142857142856 



The heuristic tells us the order:

- Biscuit 3 > Biscuit 1 > Biscuit 0 > Biscuit 2

In [ ]:
def greedy_biscuit_placement(dough, biscuits, heuristic='value'):
    """
    Place biscuits on the dough using a greedy approach based on a specified heuristic.

    Parameters:
    - dough: The Dough object representing the dough roll.
    - biscuits: A dictionary of Biscuit objects indexed by their type.
    - heuristic: The heuristic to use for sorting biscuits. Options are:
        - 'value': Sort biscuits by their value in descending order.
        - 'valength': Sort biscuits by their value-to-length ratio in descending order.
        - Other: Sort biscuits by value divided by (length * sum of max defects), in descending order.

    Returns:
    - total_value: The total value of the biscuits placed, minus the penalty for unused dough.
    - placed_biscuits: A list of tuples representing the placed biscuits as (position, biscuit_type).
    """
    # Sort biscuits based on the selected heuristic
    if heuristic == 'value':
        sorted_biscuits = sorted(biscuits.values(), key=lambda b: b.value, reverse=True)
    elif heuristic == 'valength':
        sorted_biscuits = sorted(biscuits.values(), key=lambda b: b.value / b.length, reverse=True)
    else:
        sorted_biscuits = sorted(
            biscuits.values(),
            key=lambda b: b.value / (b.length * sum(b.max_defects.values())),
            reverse=True
        )

    total_value = 0
    current_position = 0
    placed_biscuits = []
    unused_positions = set(range(dough.LENGTH))  # Set of all positions on the dough

    while current_position < dough.LENGTH:
        placed = False  # Flag to check if a biscuit has been placed at the current position
        for biscuit in sorted_biscuits:
            # Check if the biscuit fits within the remaining dough length
            if current_position + biscuit.length <= dough.LENGTH:
                # Count defects in the current segment of the dough
                defects = dough.count_defects(current_position, biscuit.length)
                # Check if defects are within the biscuit's maximum allowed defects
                if all(defects.get(cls, 0) <= biscuit.max_defects[cls] for cls in biscuit.max_defects):
                    placed_biscuits.append((current_position, biscuit.biscuit_type))
                    total_value += biscuit.value  # Add the biscuit's value
                    for pos in range(current_position, current_position + biscuit.length):
                        unused_positions.discard(pos)
                    # Move current position forward by the biscuit's length
                    current_position += biscuit.length
                    placed = True
                    break 
        if not placed:
            # No biscuit could be placed; move to the next position
            current_position += 1
            unused_positions.discard(current_position - 1)  # Mark this position as unused

    # Calculate the penalty for unused positions
    penalty = -len(unused_positions)
    total_value += penalty

    return total_value, placed_biscuits


In [22]:
def main(heuristic):
    dough = Dough(length=500)
    for _, row in df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
        
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    total_value, placed_biscuits = greedy_biscuit_placement(dough, biscuits,heuristic)
    print("Total Value:", total_value)
    print("Placed Biscuits:", placed_biscuits)

In [23]:
if __name__ == "__main__":
    start = time.time()
    main('value')
    print("solved in : ", time.time()-start,"seconds")

Total Value: 675
Placed Biscuits: [(0, 1), (8, 1), (16, 1), (24, 1), (32, 2), (34, 2), (36, 0), (40, 1), (48, 2), (50, 0), (54, 1), (62, 1), (70, 1), (78, 1), (86, 1), (95, 2), (98, 1), (106, 0), (111, 0), (118, 0), (122, 1), (130, 1), (138, 0), (142, 0), (146, 1), (154, 2), (158, 1), (167, 1), (175, 1), (185, 2), (187, 1), (195, 1), (203, 1), (211, 1), (219, 1), (227, 1), (235, 3), (242, 1), (250, 1), (258, 1), (266, 1), (274, 0), (278, 1), (286, 1), (294, 1), (302, 1), (310, 1), (318, 1), (326, 1), (334, 1), (342, 1), (351, 2), (353, 0), (357, 1), (365, 0), (369, 0), (373, 1), (381, 2), (383, 1), (391, 1), (399, 1), (407, 1), (415, 1), (423, 1), (431, 0), (435, 1), (443, 1), (451, 1), (459, 1), (467, 1), (475, 0), (479, 1), (487, 1), (495, 3)]
solved in :  0.029131650924682617 seconds


In [24]:
if __name__ == "__main__":
    start = time.time()
    main('valength')
    print("solved in : ", time.time()-start,"seconds")

Total Value: 690
Placed Biscuits: [(0, 1), (8, 1), (16, 3), (21, 1), (29, 3), (34, 2), (36, 0), (40, 1), (48, 2), (50, 0), (54, 3), (59, 3), (64, 1), (72, 1), (80, 1), (88, 3), (93, 0), (98, 1), (106, 0), (111, 0), (118, 0), (122, 1), (130, 3), (135, 0), (139, 0), (143, 0), (147, 3), (152, 3), (158, 3), (164, 2), (167, 3), (172, 3), (177, 0), (181, 2), (185, 2), (187, 1), (195, 3), (200, 3), (205, 3), (210, 3), (215, 3), (220, 3), (225, 3), (230, 3), (235, 3), (242, 1), (250, 3), (255, 3), (260, 3), (265, 3), (270, 3), (275, 0), (279, 1), (287, 3), (293, 1), (301, 1), (309, 3), (314, 1), (322, 3), (327, 3), (332, 2), (334, 3), (339, 3), (344, 3), (349, 3), (354, 1), (362, 0), (366, 0), (370, 0), (374, 1), (382, 1), (390, 3), (395, 3), (400, 1), (408, 3), (413, 3), (418, 3), (423, 1), (431, 0), (435, 1), (443, 1), (451, 3), (456, 3), (461, 3), (466, 3), (471, 3), (476, 0), (480, 1), (488, 3), (493, 3)]
solved in :  0.031054973602294922 seconds


In [25]:
if __name__ == "__main__":
    start = time.time()
    main('valengthdef')
    print("solved in : ", time.time()-start,"seconds")

Total Value: 672
Placed Biscuits: [(0, 1), (8, 1), (16, 3), (21, 2), (23, 1), (31, 3), (36, 2), (40, 1), (48, 2), (50, 0), (54, 3), (59, 3), (64, 1), (72, 2), (74, 1), (82, 1), (90, 2), (92, 0), (98, 1), (106, 2), (108, 3), (114, 2), (118, 0), (122, 2), (124, 1), (132, 3), (137, 2), (139, 2), (141, 0), (145, 1), (153, 2), (155, 2), (158, 3), (164, 2), (167, 3), (172, 3), (177, 0), (181, 2), (185, 2), (187, 1), (195, 3), (200, 3), (205, 3), (210, 3), (215, 3), (220, 3), (225, 3), (230, 3), (235, 3), (242, 1), (250, 3), (255, 3), (260, 3), (265, 3), (270, 3), (275, 0), (279, 1), (287, 3), (293, 1), (301, 1), (309, 3), (314, 1), (322, 3), (327, 3), (332, 2), (334, 3), (339, 3), (344, 3), (349, 3), (354, 1), (362, 2), (364, 0), (368, 0), (373, 1), (381, 2), (383, 3), (388, 3), (393, 3), (398, 2), (400, 1), (408, 3), (413, 3), (418, 3), (423, 2), (425, 1), (433, 1), (441, 3), (446, 1), (454, 3), (459, 3), (464, 3), (469, 3), (474, 2), (476, 2), (478, 0), (483, 1), (491, 3), (496, 0)]
solved

The greedy approach uses a straightforward, heuristic-based strategy to place biscuits on the dough. By sorting biscuits based on specific criteria, the algorithm iteratively selects the best-fit biscuit for the current position.

We obtained the following results:

- Value heuristic (main('value')) prioritizes biscuits with the highest value. This achieved a total score of 675, but it may neglect the space or defect constraints for more optimal placements.
- Value-to-length ratio heuristic (main('valength')) optimizes placements by considering the value of a biscuit relative to its length. This approach resulted in a better score of 690, as it balances space utilization and value.
- Comprehensive heuristic (main('valengthdef')) takes into account value, length, and defect tolerances. With a total score of 672, it highlights the trade-off between placement feasibility and maximizing value.

The heuristic choice influences how the dough is utilized. The value-to-length ratio heuristic stands out as a good compromise for this problem, achieving the best total score far away from the best GA (640).

# 3. CSP

## 3.1 Import CSP librairies

In [26]:
from ortools.sat.python import cp_model
model = cp_model.CpModel()

## 3.2 Create variables

In [ ]:
# Define the Biscuit types
biscuit_types = [Biscuit(i) for i in range(4)]  # Assuming biscuit types are 0 to 4

for biscuit in biscuit_types:
    print(biscuit)

# Determine the maximum starting position for biscuits, this ensures that even the largest biscuit can fit within the dough length
max_biscuit_length = max(biscuit.length for biscuit in biscuit_types)
max_start_pos = dough.LENGTH - max_biscuit_length

# Create variables for the CSP model
biscuit_vars = {}
for biscuit in biscuit_types:
    for start_pos in range(max_start_pos + 1):
        var_name = f"biscuit_{biscuit.biscuit_type}_start_{start_pos}"
        biscuit_vars[(biscuit.biscuit_type, start_pos)] = model.NewBoolVar(var_name)


Biscuit Type: 0, Length: 4, Value: 3, Max Defects: {'a': 4, 'b': 2, 'c': 3}
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}
Biscuit Type: 2, Length: 2, Value: 1, Max Defects: {'a': 1, 'b': 2, 'c': 1}
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}


In [28]:
print(biscuit_vars[(0,3)])

biscuit_0_start_3


## 3.3 Create Constraints

In [ ]:
# Constraint: No overlapping biscuits can be placed on the dough

for position in range(dough.LENGTH):
    # List to collect variables of biscuits covering the current position
    overlapping_biscuits = []
    
    for biscuit in biscuit_types:
        # Calculate possible start positions for this biscuit type that cover the current position
        start_positions = range(
            max(0, position - biscuit.length + 1),
            min(position + 1, max_start_pos + 1)
        )
        
        for start_pos in start_positions:
            # Check if the biscuit starting at 'start_pos' covers the 'position'
            if start_pos <= position < start_pos + biscuit.length:
                overlapping_biscuits.append(biscuit_vars[(biscuit.biscuit_type, start_pos)])
    
    if overlapping_biscuits:
        # At most one biscuit can cover each position
        model.Add(sum(overlapping_biscuits) <= 1)


# Constraint: Defects within each biscuit must not exceed the maximum allowed for that biscuit type

for biscuit in biscuit_types:
    for start_pos in range(max_start_pos + 1):
        for defect_type, max_allowed in biscuit.max_defects.items():
            # List to hold positions of defects of this type within the biscuit's range
            defect_positions = []
            
            for defect_position, defect_class in dough.defects_list:
                if defect_class == defect_type and start_pos <= defect_position < start_pos + biscuit.length:
                    defect_positions.append(defect_position)
            
            num_defects = len(defect_positions)
            
            # Add the constraint: If the biscuit is placed, the number of defects must not exceed the maximum allowed
            model.Add(num_defects * biscuit_vars[(biscuit.biscuit_type, start_pos)] <= max_allowed)


## 3.4 Objective function to maximize

In [30]:
# Create variables indicating whether each position on the dough is covered by a biscuit
position_covered = {}
for x in range(dough.LENGTH):
    position_covered[x] = model.NewBoolVar(f'position_covered_{x}')

# Link the position_covered variables with the biscuit placement variables
for x in range(dough.LENGTH):
    # List to collect variables of biscuits covering the current position
    covering_biscuits = []
    for biscuit in biscuit_types:
        # Calculate possible start positions for this biscuit type that cover the current position
        start_positions = range(
            max(0, x - biscuit.length + 1),
            min(x + 1, max_start_pos + 1)
        )
        for start_pos in start_positions:
            # Check if the biscuit starting at 'start_pos' covers the 'position'
            if start_pos <= x < start_pos + biscuit.length:
                covering_biscuits.append(biscuit_vars[(biscuit.biscuit_type, start_pos)])
    # Since at most one biscuit can cover each position, the sum of covering_biscuits is 0 or 1
    # Set position_covered[x] to be equal to this sum
    model.Add(position_covered[x] == sum(covering_biscuits))

# Calculate the total value of placed biscuits
total_biscuit_value = sum(
    biscuit.value * biscuit_vars[(biscuit.biscuit_type, start_pos)]
    for biscuit in biscuit_types
    for start_pos in range(max_start_pos + 1)
)

# Calculate the total number of positions covered by biscuits
total_positions_covered = sum(position_covered[x] for x in range(dough.LENGTH))

# Calculate the total penalty for unused positions
# Each unused position incurs a penalty of -1
# Total penalty = (dough.LENGTH - total_positions_covered)
total_penalty = dough.LENGTH - total_positions_covered

# Define the objective function to maximize
# Total value = total_biscuit_value - total_penalty
# Simplify the expression for better solver performance
total_value = total_biscuit_value - total_penalty

# Set the objective in the model
model.Maximize(total_value)


## 3.5 Solve the problem with CSP

In [31]:
start =time.time()
solver = cp_model.CpSolver()
status = solver.Solve(model)
print("solved in : ", time.time()-start,"seconds")

solved in :  0.26720213890075684 seconds


In [32]:
if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
    if status==cp_model.OPTIMAL:
        print("The solution found is optimal !")
    if status==cp_model.FEASIBLE:
        print("The solution found is feasible !")
solution = []
for biscuit in biscuit_types:
    for start_pos in range(max_start_pos + 1):
        if solver.Value(biscuit_vars[(biscuit.biscuit_type, start_pos)]) == 1:#if 1/True this mean that it has been placed
            solution.append((biscuit.biscuit_type, start_pos))
print("The objective function value found is : ",solver.ObjectiveValue()," and the solution found is : ",solution)

The solution found is optimal !
The objective function value found is :  715.0  and the solution found is :  [(0, 36), (0, 50), (0, 67), (0, 114), (0, 140), (0, 353), (0, 365), (0, 371), (1, 0), (1, 8), (1, 23), (1, 40), (1, 59), (1, 71), (1, 89), (1, 98), (1, 119), (1, 132), (1, 149), (1, 158), (1, 167), (1, 176), (1, 187), (1, 242), (1, 270), (1, 278), (1, 286), (1, 294), (1, 312), (1, 325), (1, 357), (1, 375), (1, 398), (1, 406), (1, 424), (1, 432), (1, 440), (1, 453), (1, 471), (1, 479), (1, 492), (2, 21), (2, 48), (2, 106), (2, 185), (2, 215), (2, 369), (3, 16), (3, 31), (3, 54), (3, 79), (3, 84), (3, 108), (3, 127), (3, 144), (3, 195), (3, 200), (3, 205), (3, 210), (3, 217), (3, 222), (3, 227), (3, 232), (3, 237), (3, 250), (3, 255), (3, 260), (3, 265), (3, 302), (3, 307), (3, 320), (3, 333), (3, 338), (3, 343), (3, 348), (3, 383), (3, 388), (3, 393), (3, 414), (3, 419), (3, 448), (3, 461), (3, 466), (3, 487)]


In [33]:
sorted_list = sorted(solution, key=lambda x: x[1])
print(sorted_list)

[(1, 0), (1, 8), (3, 16), (2, 21), (1, 23), (3, 31), (0, 36), (1, 40), (2, 48), (0, 50), (3, 54), (1, 59), (0, 67), (1, 71), (3, 79), (3, 84), (1, 89), (1, 98), (2, 106), (3, 108), (0, 114), (1, 119), (3, 127), (1, 132), (0, 140), (3, 144), (1, 149), (1, 158), (1, 167), (1, 176), (2, 185), (1, 187), (3, 195), (3, 200), (3, 205), (3, 210), (2, 215), (3, 217), (3, 222), (3, 227), (3, 232), (3, 237), (1, 242), (3, 250), (3, 255), (3, 260), (3, 265), (1, 270), (1, 278), (1, 286), (1, 294), (3, 302), (3, 307), (1, 312), (3, 320), (1, 325), (3, 333), (3, 338), (3, 343), (3, 348), (0, 353), (1, 357), (0, 365), (2, 369), (0, 371), (1, 375), (3, 383), (3, 388), (3, 393), (1, 398), (1, 406), (3, 414), (3, 419), (1, 424), (1, 432), (1, 440), (3, 448), (1, 453), (3, 461), (3, 466), (1, 471), (1, 479), (3, 487), (1, 492)]


Using a CSP approach allows us to find the optimal solution with a total value of 715, which serves as a benchmark to compare against other heuristic-based solutions. This ensures we have a reference point for evaluating the performance and quality of alternative methods like greedy or genetic algorithms.

# 4. Dynamic programming approach

## 4.1 Initialization of Dough Defects, Biscuit Types, and Dynamic Programming Structure







In [ ]:
dough_defects = [[] for _ in range(dough.LENGTH)]

for _, row in df.iterrows():
    position = int(row['x']) 
    defect_class = row['class']  
    dough_defects[position].append(defect_class)  

biscuit_types = [Biscuit(i) for i in range(4)]

maxAtindex = [[] for _ in range(500)]

## 4.2 Creating usefull functions

In [ ]:
def get_total_value(roll, dough_length):
    """
    Calculate the total value of a given arrangement of biscuits, including penalties for unused positions.

    Parameters:
    - roll (list): List of tuples containing (Biscuit, position).
    - dough_length (int): The total length of the dough.

    Returns:
    - int: Total value of all biscuits in the arrangement minus penalty for unused positions.
    """
    total_value = 0
    occupied_positions = set()

    for biscuit, position in roll:
        total_value += biscuit.value
        occupied_positions.update(range(position, position + biscuit.length))

    # Calculate penalty for unused positions
    total_positions = set(range(dough_length))
    unused_positions = total_positions - occupied_positions
    penalty = -len(unused_positions)
    total_value += penalty

    return total_value

def can_be_placed(biscuit, start_index, dough_defects):
    """
    Check if a biscuit can be placed at a given index on the dough.

    Parameters:
    - biscuit (Biscuit): The Biscuit object to be placed.
    - start_index (int): The starting index on the dough.
    - dough_defects (list): List representing defects at each position on the dough.

    Returns:
    - bool: True if the biscuit can be placed, False otherwise.
    """
    end_index = start_index + biscuit.length
    if end_index > len(dough_defects):
        return False

    defect_counts = {'a': 0, 'b': 0, 'c': 0}
    for index in range(start_index, end_index):
        defects_at_pos = dough_defects[index]
        for defect_type in defects_at_pos:
            defect_counts[defect_type] += 1

    # Check if defects are within acceptable limits
    for defect_type, max_allowed in biscuit.max_defects.items():
        if defect_counts.get(defect_type, 0) > max_allowed:
            return False

    return True


## 4.3 Dynamic programming resolution

In [ ]:
start_time = time.time()

dough_length = dough.LENGTH

max_at_index = [None] * (dough_length + 1)
max_at_index[0] = []

for position in range(1, dough_length + 1):
    max_value = float('-inf') 
    best_roll = None 

    for biscuit in biscuit_types:
        biscuit_length = biscuit.length

        if position >= biscuit_length:
            previous_roll = max_at_index[position - biscuit_length]
            if previous_roll is not None:
                current_value = get_total_value(previous_roll, position - biscuit_length)

                # Check if the biscuit can be placed at the starting position
                start_pos = position - biscuit_length
                if can_be_placed(biscuit, start_pos, dough_defects):
                    current_value += biscuit.value 

                    occupied_positions = set()
                    for placed_biscuit, pos in previous_roll:
                        occupied_positions.update(range(pos, pos + placed_biscuit.length))
                    occupied_positions.update(range(start_pos, position))

                    # Calculate the penalty for unused positions
                    total_positions = set(range(position))
                    unused_positions = total_positions - occupied_positions
                    penalty = -len(unused_positions)
                    total_current_value = current_value + penalty 

                    # Update the best sequence if the current value is higher
                    if total_current_value > max_value:
                        best_roll = previous_roll + [(biscuit, start_pos)]
                        max_value = total_current_value

    # If no sequence was found, keep the previous sequence
    if best_roll is None:
        max_at_index[position] = max_at_index[position - 1]
    else:
        # Update the best sequence at the current position
        max_at_index[position] = best_roll

# Get the best sequence at the end of the dough
best_solution = max_at_index[dough_length]
total_value = get_total_value(best_solution, dough_length)

end_time = time.time()

print(f"Found in: {end_time - start_time} seconds")
print(f"Maximum value: {total_value}")
for biscuit, position in best_solution:
    print(f"Biscuit of type {biscuit.biscuit_type} at position: {position}")


Found in: 0.10386848449707031 seconds
Maximum value: 715
Biscuit of type 1 at position: 0
Biscuit of type 1 at position: 8
Biscuit of type 3 at position: 16
Biscuit of type 1 at position: 21
Biscuit of type 3 at position: 29
Biscuit of type 2 at position: 34
Biscuit of type 0 at position: 36
Biscuit of type 1 at position: 40
Biscuit of type 2 at position: 48
Biscuit of type 0 at position: 50
Biscuit of type 3 at position: 54
Biscuit of type 3 at position: 59
Biscuit of type 1 at position: 64
Biscuit of type 1 at position: 72
Biscuit of type 1 at position: 80
Biscuit of type 3 at position: 88
Biscuit of type 0 at position: 93
Biscuit of type 1 at position: 98
Biscuit of type 2 at position: 106
Biscuit of type 3 at position: 108
Biscuit of type 0 at position: 114
Biscuit of type 1 at position: 119
Biscuit of type 3 at position: 127
Biscuit of type 1 at position: 132
Biscuit of type 0 at position: 140
Biscuit of type 3 at position: 144
Biscuit of type 1 at position: 149
Biscuit of type 1 

To go further : the Dynamic Programming approach is used to explore all possible placements of biscuits on the dough while ensuring constraints are respected. By iteratively building the optimal solution at each position based on previous placements, this method systematically finds the arrangement that maximizes the total value, achieving a score of 715, identical to the CSP benchmark. This confirms the optimality of the solution.

### Conclusion & Results

Greedy Approach:

- Best Score: 690 (using the value/length heuristic).
- Reason: The heuristic is highly effective as it prioritizes biscuits that maximize value while minimizing space usage, leading to efficient utilization of the dough. This simplicity and relevance make it surprisingly competitive with more advanced methods.

Genetic Algorithms :

- Best Score: 640 (with uniform crossover and tournament selection).
- Reason: While GAs explore the solution space through mutation and crossover, they are more stochastic and less directed compared to a tailored heuristic. This can limit their ability to consistently converge to the highest scores within a fixed number of generations.

CSP:

- Score: 715.
- Reason: CSP systematically explores all possible solutions, respecting all constraints, to find the optimal configuration. It provides a benchmark for comparing other methods but is computationally expensive.

Dynamic Programming :

- Score: 715.
- Reason: Similar to CSP, DP finds the optimal solution by breaking the problem into subproblems and solving them iteratively. It is efficient and guarantees the optimal result.


Why the Greedy Method Outperformed GA ?

The greedy method performs better than GA because of the strength and relevance of the heuristic (value/length). This heuristic directly aligns with the goal of maximizing value while minimizing space and defect penalties. By prioritizing high-value, space-efficient biscuits, the greedy algorithm achieves near-optimal results with far less computational complexity compared to the stochastic nature of genetic algorithms, which can struggle to match the precision of a well-chosen heuristic within limited iterations.

